serving/frontend/src/Components/Forms/FileForm.jsx

```diff
@@ -0,0 +1,104 @@
+import {useState} from "react";
+import {displaySubmitInfo, renderHeader} from "../../utils";
+import {minimumPropertyValues} from "./rules";
+
+const FileForm = (setMonsterProperties) => {
+    const [selectedFile, setSelectedFile] = useState(null);
+    const [selectedFileName, setSelectedFileName] = useState("");
+    const systemProperties = new Map([
+        ["abilities", ["cha", "con", "dex", "int", "str", "wis"]],  // mod
+        ["attributes", ["ac", "hp"]]  // value
+    ]);
+    const propertiesValuesKey = new Map([
+        ["abilities", "mod"],
+        ["attributes", "value"]
+    ]);
+
+    const uploadFile = (file) => {
+        if (file === undefined) {
+            return;
+        }
+        setSelectedFile(file);
+        setSelectedFileName(file.name);
+    };
+
+    const unpackValue = (dict, dictKeys) => {
+        let current = dict;
+        for (let dictKey of dictKeys) {
+            if (! current.hasOwnProperty(dictKey)) {
+                const keyPath = dictKeys.join("/");
+                throw new Error("Selected JSON is invalid: value from " + keyPath + " was not found.")
+            }
+            current = current[dictKey];
+        }
+        return current;
+    }
+
+    const parseFile = (fileReader) => {
+        const fileDict = JSON.parse(fileReader.result);
+        const systemDict = fileDict.system;
+        let resultDict = {};
+        try {
+            resultDict["name"] = fileDict.name;
+            systemProperties.forEach((subproperties, property) => {
+                const valuesKey = propertiesValuesKey.get(property);
+                for (let subproperty of subproperties) {
+                    const unpackedValue = unpackValue(systemDict, [property, subproperty, valuesKey]);
+                    resultDict[subproperty] = unpackedValue;
+                    if (unpackedValue < minimumPropertyValues.get(subproperty)) {
+                        throw new Error("Selected JSON is invalid: value of " + property + "/" + subproperty +
+                            " has to be grater than or equal to " + minimumPropertyValues.get(subproperty) +
+                            " (currently is " + unpackedValue + ").");
+                    }
+                }
+            })
+        } catch (e) {
+            alert(e);
+            return null;
+        }
+        return resultDict;
+    };
+
+    const submitForm = async (e) => {
+        e.preventDefault();
+        let reader = new FileReader();
+        reader.readAsText(selectedFile);
+        reader.addEventListener("load", (event) => {
+            const properties = parseFile(reader);
+            if (properties === null) {
+                return;
+            }
+            fetch("http://localhost:8000/properties/upload", {
```

Comment:
Proponuję wyeksportować adres backendu i wczytywać to ze zmiennej środowiskowej, bo nie ma gwarancji, że to jest localhost

Reply:
Wyeksportować czyli zapisać go w dockerfile'u?

Reply:
Najprościej w JSie jest użyć dotenv: https://zetcode.com/javascript/dotenv/. Bo to lokalnie do testów jest localhost, ale jak to wdrożycie na chmurę, no to już będzie inny adres, stąd problem.

Reply:
A, i ten endpoint /upload w ogóle już jest zaorany na backendzie, nie? To trzeba by zaktualizować na froncie te URLe. No i jak to jest w jednym .env, to się zmienia tylko w jednym miejscu.

Reply:
> A, i ten endpoint /upload w ogóle już jest zaorany na backendzie, nie? To trzeba by zaktualizować na froncie te URLe. No i jak to jest w jednym .env, to się zmienia tylko w jednym miejscu.

Endpoint jest poprawny.